packages/core/saved-objects/core-saved-objects-migration-server-internal/src/actions/delete_by_query.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Either from 'fp-ts/lib/Either';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import type {
+  BulkIndexByScrollFailure,
+  DeleteByQueryResponse,
+  QueryDslQueryContainer,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import {
+  catchRetryableEsClientErrors,
+  type RetryableEsClientError,
+} from './catch_retryable_es_client_errors';
+
+/** @internal */
+export interface DeleteByQueryParams {
+  client: ElasticsearchClient;
+  indexName: string;
+  query: QueryDslQueryContainer;
+}
+
+/** @internal */
+export interface DeleteByQueryErrorResponse {
+  type: 'delete_failed';
+  conflictingDocuments: BulkIndexByScrollFailure[];
+}
+
+/**
+ * Deletes documents matching the provided query
+ */
+export const deleteByQuery = ({
+  client,
+  indexName,
+  query,
+}: DeleteByQueryParams): TaskEither.TaskEither<
+  RetryableEsClientError | DeleteByQueryErrorResponse,
+  'delete_successful'
+> => {
+  return () => {
+    return client
+      .deleteByQuery({
+        index: indexName,
+        query,
+        wait_for_completion: true,
```

Comment:
this would mean the requests fails if it takes longer than the default timeout to complete. I think we set that to 120s but it's probably worth validating that when a deleteByQuery takes too long we get a retryable_es_client error. Because this could easily happen but as long as we retry it should be fine.

There is a risk that we create a new deleteByQuery before the existing one completed, but with our exponential backoff that should be fine. Also don't think we should usually have a huge number of docs deleted.

Reply:
Fair point, thanks for the insight!

I take it we still need that flag, cause not adding it and carrying on with the flow could have worse consequences.

I'll validate the timeout doubt.

Reply:
yes, without it we'd get back a task and we'd need to add a step like the other `*_WAIT_FOR_TASK` steps. I didn't want to have to add another step but does actually feel safer. Otherwise if a cluster is slow we could have several Kibana's piling on several deleteByQueries on top of each other making matters worse.

Delete should be quicker than the updateByQuery we run in `UPDATE_TARGET_MAPPINGS` but for that action we did have this problem a few times in production.

So it's starting to feel like it'd be safer to just create a task and wait for it and know we don't have anything to worry about.